### ISSUE TITLE
Create Jax Interface

### ISSUE THREAD

**Issue by Thenerdstation** [PennyLaneAI/pennylane#943]

**Comment by Thenerdstation** [comment]
I now see why we haven't done this already, as it's not straightforward.

The problem is that Jax has no way to wrap pure numpy function (`numpy -> numpy`) to a jax function (`jax -> numpy -> numpy -> jax`). This is not a problem for Torch and TF, as they have builtin ways to do this (torch.autograd.Function and eagermode/tf.py_func respectively).

However, there is a workaround, but it's not clean. XLA (which is what jax compiles to) has a [CustomCall method](https://github.com/tensorflow/tensorflow/blob/acb84a010d1859b15fc21b64978faf20bd62956e/tensorflow/compiler/xla/g3doc/custom_call.md) that lets you run mostly arbitrary C++.  What we could do is then have the [C++ call the respective python code](https://www.codeproject.com/Articles/11805/Embedding-Python-in-C-C-Part-I). Raw C++ integration seems like it would be super messy, so I'm currently looking into cython integration instead, which lets us call pure python code much more easily.

Thoughts? We already want to add C++/cython with the lighting qubit integration, so I don't think this plan would significantly complicate things beyond what we already have planned.

**Comment by Thenerdstation** [comment]
Of course, if we finish #942 and make a jax device, this doesn't apply.

**Comment by josh146** [comment]
> The problem is that Jax has no way to wrap pure numpy function (`numpy -> numpy`) to a jax function (`jax -> numpy -> numpy -> jax`). This is not a problem for Torch and TF, as they have builtin ways to do this (torch.autograd.Function and eagermode/tf.py_func respectively).

Yep, this is the problem I ran into earlier iirc

> However, there is a workaround, but it's not clean. XLA (which is what jax compiles to) has a CustomCall method that lets you run mostly arbitrary C++. What we could do is then have the C++ call the respective python code.

Yes this definitely doesn't seem clean 😆

Is this the thinking?

```
jax --> classical component compiled to XLA
    --> CustomCall method implemented in Cython
    --> PennyLane tape/devices in Python
    --> back to C++
    --> back to jax
```

> Of course, if we finish #942 and make a jax device, this doesn't apply.

With a jax backend, you will be able to do end-to-end jax computations, which is awesome. The interface is almost orthogonal; it allows you to take a hybrid quantum-classical model written in jax and allow it to run on _non_ jax devices.

So I suppose we have several options:

* **Jax device only**: the classical part of the model can use jax, but only if using the jax device. Only standard backprop is supported, no adjoint/parameter-shift. This also means you cannot train with quantum devices that include stochasticity (e.g., by specifying a finite number of shots).

* **Jax device, and Jax interface that can be used with lightning qubit**: The jax device allows for end-to-end jax backprop as above, but we include a Jax interface that is tightly coupled to lightning.qubit. This allows for using jax to train a model with parameter-shift, finite shots, etc., but you must use lightning.qubit -- you cannot swap to other devices (including hardware).

* **Jax device, and general jax interface**: the most general solution, allows for complete plug and play between devices without needing to give up jax for the classical part of the model. However, subject to the caveats you mention above.

**Comment by Thenerdstation** [comment]
> Yes this definitely doesn't seem clean 😆

Yeah its solidly in the "lets not do that" regime. 

I think the first two suggestions you listed are the correct options. 

> Only standard backprop is supported, no adjoint/parameter-shift.

Why would this be the case? Assuming our jax device has access to the full wavefunction, we can get gradients however we want. 

>  But it will be hard to communicate to users the devices that work with jax, and those that don't.

Yeah, that's a concern I have as well. I think the best option here is to throw a clear error for devices that aren't yet supported.

**Comment by josh146** [comment]
> Why would this be the case? Assuming our jax device has access to the full wavefunction, we can get gradients however we want.

Oh that's a good point! I hadn't considered that. By "default" Jax classical ML + Jax device will be backprop, but I suppose that jax gradient wrapping _can_ work find with a purely jax backend.

I guess it's then a UI/implementation question - how would we signify to jax to 'change' the gradient computation and call our custom implementation? That part is not 100% clear to me

**Comment by Thenerdstation** [comment]
> how would we signify to jax to 'change' the gradient computation and call our custom implementation?

With functional programming! :)

Basically we just [wrap the forward pass function in a `custom_vjp`,](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html) and just dynamically swap which the reverse mode function we use. How the API of that should look like from the users perspective is up to you guys.
